python/revng/api/listenable_manager.py

```diff
@@ -0,0 +1,60 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import sys
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+from enum import Enum, auto
+from typing import Any, Callable, Dict, List
+
+from .manager import Manager
+
+
+class EventType(Enum):
+    BEGIN = auto()
+    CONTEXT = auto()
+
+
+EventHandler = Callable[[Manager, EventType], Any]
+
+
+class ListenableManager(Manager):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._listeners: Dict[EventType, List[EventHandler]] = defaultdict(list)
+        self._executor = ThreadPoolExecutor()
+
+    def add_event_listener(self, type_: EventType, callback: EventHandler):
+        self._listeners[type_].append(callback)
+
+    def handle_event(self, type_: EventType):
+        def handler():
+            if type_ in self._listeners:
+                for listener in self._listeners[type_]:
+                    try:
+                        listener(self, type_)
+                    except Exception as e:
+                        print(
+                            f"Caught exception {repr(e)} in running listener callback",
+                            file=sys.stderr,
+                            flush=True,
+                        )
```

Comment:
The issue here is that the executor will swallow up all exceptions, so if there is an exception it will not be shown and the only way to capture it is to do `.result()` on the future returned by the submit. Stopping on exception can be achieved via coroutines but this would complicate the structure of the program.

Reply:
Checked around, it's not handled correctly via a callback.
If there is an exception it will `sys.exit(1)`.

Reply:
Double checked, the callback is also run within the thread.
The only way (without writing a lot more scaffolding) is to issue a `SIGINT` on the process.
The traceback is now properly printed with the logger before exiting.